### PR TITLE
LDS implementation

### PR DESCRIPTION
**Pull request by kyessenov**
Description in the team document https://docs.google.com/document/d/1Xj51lucolqT9y4d9QLlhRWQfbOMyBCNCVziRfbO6n00/edit#heading=h.8qvii5zwgqk

### CODE REVIEW

**Comment by rshriram** [comment]
This is too big Kuat. I know most of it is golden files. One suggestion: how about we push all the .golden files in as one PR (first one), the subsequent PR deals only with Go code, making it easier to review. (It's fine if we have redundant golden files - we can always clean it up later. ). Otherwise it's going to be a pain to review this and follow up on the comments.

**Comment by kyessenov** [comment]
jenkins rebuild pilot/e2e-smoketest

File: proxy/envoy/discovery.go [hunk 1]
```diff
@@ -256,12 +256,6 @@ func (ds *DiscoveryService) Register(container *restful.Container) {
 		Doc("SDS registration").
 		Param(ws.PathParameter(ServiceKey, "tuple of service name and tag name").DataType("string")))
 
-	// List all known clusters (informational, not invoked by Envoy)
-	ws.Route(ws.
-		GET("/v1/clusters").
-		To(ds.ListAllClusters).
-		Doc("Clusters in CDS"))
```
Comment (rshriram) [inline proxy/envoy/discovery.go]: we still have the /v1/all api right? to get entire config dump?

File: proxy/envoy/resources.go [hunk 2]
```diff
@@ -42,6 +45,9 @@ const (
 	// CDSName is the name of CDS cluster
 	CDSName = "cds"
 
+	// RedirectListenerName is the name for traffic capture listener
+	RedirectListenerName = "redirect"
+
```
Comment (rshriram) [inline proxy/envoy/resources.go]: nit: Lets call it VirtualListenerName .. or MasterListenerName.. Redirect is bit confusing.

File: proxy/envoy/discovery.go [hunk 3]
```diff
@@ -579,64 +590,85 @@ func (ds *DiscoveryService) allServiceNodes() []string {
...
-		// apply auth policies
-		switch ds.MeshConfig.AuthPolicy {
-		case proxyconfig.ProxyMeshConfig_NONE:
-		case proxyconfig.ProxyMeshConfig_MUTUAL_TLS:
-			// apply SSL context to enable mutual TLS between Envoy proxies
-			for _, cluster := range clusters {
-				ports := model.PortList{cluster.port}.GetNames()
-				serviceAccounts := ds.Accounts.GetIstioServiceAccounts(cluster.hostname, ports)
-				cluster.SSLContext = buildClusterSSLContext(ds.MeshConfig.AuthCertsPath, serviceAccounts)
+			if cluster.port != nil {
```
Comment (rkpagadala) [inline proxy/envoy/discovery.go]: nit: reduce indentation?
if cluster.port == nil {
    continue
}
remaining code

File: proxy/context.go [hunk 4]
```diff
@@ -15,48 +15,111 @@
...
-	MeshConfig *proxyconfig.ProxyMeshConfig
+	// Mesh is the mesh config (to be merged into the config store)
+	Mesh *proxyconfig.ProxyMeshConfig
+}
+
+// Role declares the proxy node role in the mesh
+type Role interface {
+	// nolint: megacheck
+	isProxyRole()
+
```
Comment (kyessenov) [inline proxy/context.go]: that's how subtyping works in golang, using a dummy method in a marker interface.

File: proxy/context.go [hunk 5]
```diff
@@ -15,48 +15,111 @@
...
-	//
-	// The passthrough ports should be exposed irrespective of the services
-	// model. In case there is an overlap, that is the port is utilized by a
-	// service for a service instance and is also present in this list, the
-	// service model declaration takes precedence. That means any protocol
-	// upgrade (such as utilizng TLS for proxy-to-proxy traffic) will be applied
-	// to the passthrough port.
-	PassthroughPorts []int
+	// PodName for the proxy sidecar
+	PodName string
```
Comment (kyessenov) [inline proxy/context.go]: ok

File: proxy/context.go [hunk 6]
```diff
@@ -15,48 +15,111 @@
...
+
+	// PodNamespace for the proxy sidecar
+	PodNamespace string
```
Comment (kyessenov) [inline proxy/context.go]: it's to make it easy to encode/decode into service node string. there's no special meaning for it.

File: proxy/context.go [hunk 7]
```diff
@@ -15,48 +15,111 @@
...
+
+// ServiceNode for sidecar
+func (role Sidecar) ServiceNode() string {
+	return fmt.Sprintf("%s.%s.%s", role.IPAddress, role.PodName, role.PodNamespace)
+}
+
+// PodID uniquely identifies a pod
+func (role Sidecar) PodID() string {
+	return fmt.Sprintf("kubernetes://%s.%s", role.PodName, role.PodNamespace)
+}
```
Comment (kyessenov) [inline proxy/context.go]: yes, it's unfortunate but I can't change mixer code here since it reads this ID.

File: cmd/pilot/main.go [hunk 8]
```diff
@@ -256,17 +219,15 @@ func init() {
...
-		"IP address. If not provided uses ${POD_IP} environment variable.")
-	proxyCmd.PersistentFlags().StringVar(&flags.podName, "podName", "",
-		"Pod name. If not provided uses ${POD_NAME} environment variable")
-
-	sidecarCmd.PersistentFlags().IntSliceVar(&flags.passthrough, "passthrough", nil,
-		"Passthrough ports for health checks")
+	proxyCmd.PersistentFlags().StringVar(&flags.sidecar.IPAddress, "ipAddress", "",
+		"Sidecar IP address. If not provided uses ${POD_IP} environment variable.")
+	proxyCmd.PersistentFlags().StringVar(&flags.sidecar.PodName, "podName", "",
+		"Sidecar pod name. If not provided uses ${POD_NAME} environment variable")
```
Comment (kyessenov) [inline cmd/pilot/main.go]: done

File: proxy/envoy/discovery.go (hunk 1 above)
Comment (kyessenov) [inline proxy/envoy/discovery.go]: i removed clusters temporarily. the code to list all configs needs to be cleaned up but it's useful, I get it.

File: proxy/envoy/resources.go (hunk 2 above)
Comment (kyessenov) [inline proxy/envoy/resources.go]: done

File: proxy/context.go [hunk 9]
```diff
@@ -15,48 +15,111 @@
...
-	// MeshConfig defines global configuration settings
-	MeshConfig *proxyconfig.ProxyMeshConfig
+	// Mesh is the mesh config (to be merged into the config store)
+	Mesh *proxyconfig.ProxyMeshConfig
+}
+
+// Role declares the proxy node role in the mesh
+type Role interface {
+	// nolint: megacheck
+	isProxyRole()
```
Comment (kyessenov) [inline proxy/context.go]: It's a dummy method in a marker interface.

File: proxy/envoy/discovery.go [hunk 10]
```diff
@@ -579,64 +590,85 @@ func (ds *DiscoveryService) allServiceNodes() []string {
...
-		httpRouteConfigs, _ = buildIngressRoutes(ds.Config.IngressRules(), ds.Discovery, ds.Config)
-	case egressNode:
-		httpRouteConfigs = buildEgressRoutes(ds.Discovery, ds.MeshConfig)
+	case proxy.IngressNode:
+		httpRouteConfigs, _ := buildIngressRoutes(ds.IngressRules(), ds, ds)
+		clusters = httpRouteConfigs.clusters().normalize()
+
+		listener := buildHTTPListener(ds.Mesh, nil, WildcardAddress, 443, true, true)
+		listener.SSLContext = &SSLContext{
+			CertChainFile:  IngressCerts + "tls.crt",
```
Comment (kyessenov) [inline proxy/envoy/discovery.go]: done

File: proxy/envoy/discovery.go (hunk 3 above)
Comment (kyessenov) [inline proxy/envoy/discovery.go]: done. thanks!

File: proxy/envoy/resources.go [hunk 11]
```diff
@@ -472,10 +481,16 @@ type Listener struct {
 // Listeners is a collection of listeners
 type Listeners []*Listener
 
-// normalize sorts listeners by address
+// normalize sorts and de-duplicates listeners by address
 func (listeners Listeners) normalize() Listeners {
-	out := make(Listeners, len(listeners))
-	copy(out, listeners)
+	out := make(Listeners, 0)
```
Comment (kyessenov) [inline proxy/envoy/resources.go]: done. thanks!

File: proxy/context.go (hunk 6 above)
Comment (rshriram) [inline proxy/context.go]: Right but can't we concatenate the env vars and create a single podname string to be passed to proxy/ ? 
Just concerned that we are introducing minor semantics (components of pod fqdn)

**Comment by kyessenov** [comment]
I'll address the instance ID encoding in a subsequent PR. I also left out domain suffix for now, so there's definitely another PR coming.

File: proxy/envoy/config.go [hunk 12]
```diff
@@ -112,6 +118,15 @@ func buildConfig(listeners Listeners, clusters Clusters, mesh *proxyconfig.Proxy
 		StatsdUDPIPAddress: mesh.StatsdUdpAddress,
 	}
 
+	if lds {
```
Comment (andraxylia) [inline proxy/envoy/config.go]: Is there ever a case where lds is not enabled?

Comment (kyessenov) [inline proxy/envoy/config.go]: Not yet. I think Istio-on-Istio might need to skip LDS. A proxy in front of pilot might need to be statically configured.

File: proxy/envoy/discovery.go [hunk 13]
```diff
@@ -401,44 +412,20 @@ func (ds *DiscoveryService) ListEndpoints(request *restful.Request, response *re
...
-		})
-	}
-
-	sort.Slice(allClusters, func(i, j int) bool { return allClusters[i].ServiceNode < allClusters[j].ServiceNode })
-
-	if err := response.WriteEntity(allClusters); err != nil {
-		glog.Warning(err)
-	}
-}
-
```
Comment (andraxylia) [inline proxy/envoy/discovery.go]: Is thisfct ListAllClusters() no longer needed or it was never needed?

File: proxy/envoy/discovery.go [hunk 14]
```diff
@@ -523,22 +509,48 @@ func (ds *DiscoveryService) ListRoutes(request *restful.Request, response *restf
 	writeResponse(response, out)
 }
 
+// ListListeners responds to LDS requests
+func (ds *DiscoveryService) ListListeners(request *restful.Request, response *restful.Response) {
+	key := request.Request.URL.String()
+	out, cached := ds.ldsCache.cachedDiscoveryResponse(key)
+	if !cached {
```
Comment (andraxylia) [inline proxy/envoy/discovery.go]: Consider inverting the logic to reduce indentation:
if cached { 
     write 
     return 
}

// the long logic here

**Pull request merged**